File: macos_arm64/Sources/RimeoAgentMac/Services/AudioService.swift
import Foundation

// Handles waveform generation, artwork extraction, AIFF→WAV conversion
final class AudioService {
    static let shared = AudioService()
    private let convLocks = NSMapTable<NSString, NSLock>(keyOptions: .strongMemory,
                                                         valueOptions: .strongMemory)
    private let lockGuard = NSLock()

    private func convLock(for id: String) -> NSLock {
        lockGuard.lock(); defer { lockGuard.unlock() }
        let key = id as NSString
        if let existing = convLocks.object(forKey: key) { return existing }
        let lock = NSLock()
        convLocks.setObject(lock, forKey: key)
        return lock
    }

    // Returns path to a WAV file (from cache or freshly converted)
    func ensureWAV(path: String, trackID: String) throws -> String {
        let cached = AppConfig.shared.cacheDir.appendingPathComponent("conv_\(trackID).wav")
        if FileManager.default.fileExists(atPath: cached.path) {
            logger.info("AIFF conversion cache hit: track=\(trackID), wav=\(cached.path)")
            return cached.path
        }

        let lock = convLock(for: trackID)
        lock.lock(); defer { lock.unlock() }
        if FileManager.default.fileExists(atPath: cached.path) {
            logger.info("AIFF conversion cache hit after lock: track=\(trackID), wav=\(cached.path)")
            return cached.path
        }

        logger.info("Converting AIFF → WAV: \(trackID)")
        TCCDiagnostics.logPathAccess("convert-aiff", path: path)
        let exists = FileManager.default.fileExists(atPath: path)
        let readable = FileManager.default.isReadableFile(atPath: path)
        TCCDiagnostics.logPathResult("convert-aiff", path: path, exists: exists, readable: readable)
        logger.info("AIFF conversion input: track=\(trackID), exists=\(exists), readable=\(readable), path=\(path)")
        if let ffmpegPath = findBinary("ffmpeg") {
            logger.info("AIFF conversion ffmpeg: track=\(trackID), binary=\(ffmpegPath)")
        } else {
            logger.warning("AIFF conversion ffmpeg missing: track=\(trackID)")
        }
        let result = runFFmpeg(["-i", path, "-f", "wav", cached.path, "-y"], timeout: 120)
        guard result.success, FileManager.default.fileExists(atPath: cached.path) else {
            logger.warning("AIFF conversion failed: track=\(trackID), stderr=\(result.stderr)")
            throw AudioError.conversionFailed(result.stderr)
        }
        let size = (try? FileManager.default.attributesOfItem(atPath: cached.path))?[.size] as? Int ?? 0
        logger.info("AIFF conversion complete: track=\(trackID), wav=\(cached.path), bytes=\(size)")
        return cached.path
    }

    // Generate waveform JSON: {duration, peaks}
    func waveform(path: String, trackID: String) -> [String: Any] {
        let cacheURL = AppConfig.shared.cacheDir.appendingPathComponent("wave_\(trackID).json")
        if let data = try? Data(contentsOf: cacheURL),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            return json
        }

        let exists = FileManager.default.fileExists(atPath: path)
        let readable = FileManager.default.isReadableFile(atPath: path)
        TCCDiagnostics.logPathResult("waveform", path: path, exists: exists, readable: readable)
        guard exists else {
            return ["duration": 0.0, "peaks": [Double]()]
        }

        // Probe duration
        let probeResult = runFFmpeg(
            ["-v", "error", "-show_entries", "format=duration",
             "-of", "default=noprint_wrappers=1:nokey=1", path],
            binary: "ffprobe",
            timeout: 12
        )
        let duration = Double(probeResult.stdout.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0.0

        // Downsample to 8000 peaks via s8 raw PCM
        let cmd = ["-v", "error", "-i", path, "-ac", "1",
                   "-filter:a", "aresample=100", "-f", "s8", "-"]
        let result = runFFmpeg(cmd, timeout: 60)
        guard result.success, !result.rawOutput.isEmpty else {
            logger.warning("Waveform failed: track=\(trackID), stderr=\(result.stderr)")
            return ["duration": duration, "peaks": [Double]()]
        }

        let bytes  = result.rawOutput
        let samples = bytes.map { b -> Double in
            let signed = b < 128 ? Int(b) : Int(b) - 256
            return Double(abs(signed)) / 128.0
        }
        let step  = max(1, samples.count / 8000)
        var peaks = [Double]()
        var i = 0
        while i < samples.count {
            let chunk = samples[i..<min(i + step, samples.count)]
            let avg   = chunk.reduce(0, +) / Double(chunk.count)
            peaks.append(min(1.0, (avg * 2.5 * 1000).rounded() / 1000))
            i += step
        }

        let out: [String: Any] = ["duration": duration, "peaks": peaks]
        if let data = try? JSONSerialization.data(withJSONObject: out) {
            try? data.write(to: cacheURL)
        }
        return out
    }

    // Extract artwork JPEG → cache, returns path or nil
    func artwork(path: String, trackID: String) -> String? {
        let cacheURL = AppConfig.shared.cacheDir.appendingPathComponent("art_\(trackID).jpg")
        if FileManager.default.fileExists(atPath: cacheURL.path) { return cacheURL.path }
        let exists = FileManager.default.fileExists(atPath: path)
        let readable = FileManager.default.isReadableFile(atPath: path)
        TCCDiagnostics.logPathResult("artwork", path: path, exists: exists, readable: readable)
        guard exists else { return nil }

        // WAV/AIFF and many plain audio files do not contain embedded artwork streams.
        // Probe first so we do not treat a "no stream" case as an extraction failure.
        let probe = runFFmpeg(
            [
                "-v", "error",
                "-select_streams", "v:0",
                "-show_entries", "stream=index",
                "-of", "csv=p=0",
                path
            ],
            binary: "ffprobe",
            timeout: 12
        )
        let hasArtworkStream = !probe.stdout.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        if !hasArtworkStream {
            logger.info("Artwork stream missing: track=\(trackID)")
            return nil
        }

        let result = runFFmpeg([
            "-v", "error",
            "-i", path,
            "-map", "0:v:0",
            "-an",
            "-vcodec", "mjpeg",
            "-vframes", "1",
            "-s", "512x512",
            cacheURL.path,
            "-y"
        ], timeout: 45)
        let ok = result.success && FileManager.default.fileExists(atPath: cacheURL.path)
        if !ok {
            logger.warning("Artwork extraction failed: track=\(trackID), stderr=\(result.stderr)")
        }
        return ok ? cacheURL.path : nil
    }

    // Probe duration only
    func probeDuration(_ path: String) -> Double {
        TCCDiagnostics.logPathAccess("probe-duration", path: path)
        let r = runFFmpeg(
            ["-v", "error", "-show_entries", "format=duration",
             "-of", "default=noprint_wrappers=1:nokey=1", path],
            binary: "ffprobe",
            timeout: 12
        )
        return Double(r.stdout.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 300.0
    }

    // Extract audio segment to temp WAV at 22050 Hz mono
    func extractSegment(path: String, start: Double, duration: Double) -> String? {
        let tmp = FileManager.default.temporaryDirectory
            .appendingPathComponent("rimeo_seg_\(UUID().uuidString).wav")
        TCCDiagnostics.logPathAccess("extract-segment", path: path)
        let result = runFFmpeg([
            "-v", "error",
            "-ss", String(start), "-t", String(duration),
            "-i", path,
            "-ac", "1", "-ar", "22050",
            "-f", "wav", "-y", tmp.path
        ], timeout: 45)
        guard result.success, FileManager.default.fileExists(atPath: tmp.path) else {
            logger.warning("Segment extraction failed: stderr=\(result.stderr)")
            return nil
        }
        return tmp.path
    }
}

// MARK: - FFmpeg runner

struct RunResult {
    let success:   Bool
    let stdout:    String
    let stderr:    String
    let rawOutput: [UInt8]
}

func runFFmpeg(_ args: [String], binary: String = "ffmpeg", timeout: TimeInterval = 120) -> RunResult {
    guard let bin = findBinary(binary) else {
        logger.warning("Audio command failed: \(binary) not found")
        return RunResult(success: false, stdout: "", stderr: "\(binary) not found", rawOutput: [])
    }
    let proc = Process()
    proc.executableURL = URL(fileURLWithPath: bin)
    proc.arguments     = args

    let outPipe = Pipe()
    let errPipe = Pipe()
    proc.standardOutput = outPipe
    proc.standardError  = errPipe

    let outputQueue = DispatchQueue(label: "rimeo.\(binary).output", qos: .utility, attributes: .concurrent)
    let outGroup = DispatchGroup()
    var rawOut = Data()
    var rawErr = Data()

    outGroup.enter()
    outputQueue.async {
        rawOut = outPipe.fileHandleForReading.readDataToEndOfFile()
        outGroup.leave()
    }

    outGroup.enter()
    outputQueue.async {
        rawErr = errPipe.fileHandleForReading.readDataToEndOfFile()
        outGroup.leave()
    }

    do {
        try proc.run()
    } catch {
        return RunResult(success: false, stdout: "", stderr: "\(error)", rawOutput: [])
    }

    let finished = DispatchSemaphore(value: 0)
    DispatchQueue.global(qos: .utility).async {
        proc.waitUntilExit()
        finished.signal()
    }

    if finished.wait(timeout: .now() + timeout) == .timedOut {
        proc.terminate()
        _ = finished.wait(timeout: .now() + 5)
        outPipe.fileHandleForReading.closeFile()
        errPipe.fileHandleForReading.closeFile()
        _ = outGroup.wait(timeout: .now() + 5)
        return RunResult(success: false, stdout: "", stderr: "\(binary) timed out", rawOutput: [])
    }

    _ = outGroup.wait(timeout: .now() + 5)
    let stdout = String(data: rawOut, encoding: .utf8) ?? ""
    let stderr = String(data: rawErr, encoding: .utf8) ?? ""
    let raw    = [UInt8](rawOut)

    return RunResult(success: proc.terminationStatus == 0, stdout: stdout, stderr: stderr, rawOutput: raw)
}

func findBinary(_ name: String) -> String? {
    let bundled = Bundle.main.bundleURL
        .appendingPathComponent("Contents/MacOS/\(name)").path
    if FileManager.default.isExecutableFile(atPath: bundled) { return bundled }

    let component = ComponentManager.shared.componentURL(id: name).path
    if FileManager.default.isExecutableFile(atPath: component) { return component }

    return findSystemExecutable(name)
}

enum AudioError: Error {
    case conversionFailed(String)
    case fileNotFound
}

File: windows_csharp/RimeoAgent/Services/AudioService.cs
using RimeoAgent.Config;

namespace RimeoAgent.Services;

public sealed class AudioService
{
    public static readonly AudioService Shared = new();

    private readonly Dictionary<string, SemaphoreSlim> _convLocks = new();
    private readonly object _lockGuard = new();

    private SemaphoreSlim ConvLock(string id)
    {
        lock (_lockGuard)
        {
            if (!_convLocks.TryGetValue(id, out var sem))
            {
                sem = new SemaphoreSlim(1, 1);
                _convLocks[id] = sem;
            }
            return sem;
        }
    }

    public async Task<string> EnsureWav(string path, string trackId)
    {
        var cached = Path.Combine(AppConfig.Shared.CacheDir, $"conv_{trackId}.wav");
        if (File.Exists(cached)) return cached;

        var sem = ConvLock(trackId);
        await sem.WaitAsync();
        try
        {
            if (File.Exists(cached)) return cached;

            Log.Info($"Converting AIFF → WAV: {trackId}");
            var result = RunProcess("ffmpeg", new[] { "-i", path, "-f", "wav", cached, "-y" }, 120);
            if (!result.Success || !File.Exists(cached))
                throw new Exception($"AIFF conversion failed: {result.Stderr}");

            Log.Info($"AIFF conversion complete: track={trackId}, wav={cached}");
            return cached;
        }
        finally { sem.Release(); }
    }

    public Dictionary<string, object> Waveform(string path, string trackId)
    {
        var cacheFile = Path.Combine(AppConfig.Shared.CacheDir, $"wave_{trackId}.json");
        if (File.Exists(cacheFile))
        {
            try
            {
                var cached = System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, object>>(
                    File.ReadAllText(cacheFile));
                if (cached != null) return cached;
            }
            catch { }
        }

        if (!File.Exists(path))
            return new Dictionary<string, object> { ["duration"] = 0.0, ["peaks"] = Array.Empty<double>() };

        var probe = RunProcess("ffprobe", new[]
        {
            "-v", "error",
            "-show_entries", "format=duration",
            "-of", "default=noprint_wrappers=1:nokey=1",
            path
        }, 12);
        var duration = double.TryParse(probe.Stdout.Trim(), System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture, out var d) ? d : 0.0;

        var cmd = new[] { "-v", "error", "-i", path, "-ac", "1", "-filter:a", "aresample=100", "-f", "s8", "-" };
        var result = RunProcess("ffmpeg", cmd, 60);
        if (!result.Success || result.RawOutput.Length == 0)
        {
            Log.Warn($"Waveform failed: track={trackId}, stderr={result.Stderr}");
            return new Dictionary<string, object> { ["duration"] = duration, ["peaks"] = Array.Empty<double>() };
        }

        var bytes = result.RawOutput;
        var samples = bytes.Select(b => Math.Abs(b < 128 ? (int)b : (int)b - 256) / 128.0).ToArray();
        var step = Math.Max(1, samples.Length / 8000);
        var peaks = new List<double>();
        for (int i = 0; i < samples.Length; i += step)
        {
            var chunk = samples.Skip(i).Take(step).ToArray();
            var avg   = chunk.Average();
            peaks.Add(Math.Min(1.0, Math.Round(avg * 2.5 * 1000) / 1000));
        }

        var out_ = new Dictionary<string, object> { ["duration"] = duration, ["peaks"] = peaks.ToArray() };
        try
        {
            File.WriteAllText(cacheFile, System.Text.Json.JsonSerializer.Serialize(out_));
        }
        catch { }
        return out_;
    }

    public string? Artwork(string path, string trackId)
    {
        var cacheFile = Path.Combine(AppConfig.Shared.CacheDir, $"art_{trackId}.jpg");
        if (File.Exists(cacheFile)) return cacheFile;
        if (!File.Exists(path)) return null;

        var probe = RunProcess("ffprobe", new[]
        {
            "-v", "error", "-select_streams", "v:0",
            "-show_entries", "stream=index", "-of", "csv=p=0", path
        }, 12);
        if (string.IsNullOrWhiteSpace(probe.Stdout.Trim())) return null;

        var result = RunProcess("ffmpeg", new[]
        {
            "-v", "error", "-i", path, "-map", "0:v:0", "-an",
            "-vcodec", "mjpeg", "-vframes", "1", "-s", "512x512",
            cacheFile, "-y"
        }, 45);
        var ok = result.Success && File.Exists(cacheFile);
        if (!ok) Log.Warn($"Artwork extraction failed: track={trackId}");
        return ok ? cacheFile : null;
    }

    public double ProbeDuration(string path)
    {
        var r = RunProcess("ffprobe", new[]
        {
            "-v", "error", "-show_entries", "format=duration",
            "-of", "default=noprint_wrappers=1:nokey=1", path
        }, 12);
        return double.TryParse(r.Stdout.Trim(), System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture, out var d) ? d : 300.0;
    }

    public string? ExtractSegment(string path, double start, double duration)
    {
        var tmp = Path.Combine(Path.GetTempPath(), $"rimeo_seg_{Guid.NewGuid():N}.wav");
        var result = RunProcess("ffmpeg", new[]
        {
            "-v", "error",
            "-ss", start.ToString("F1", System.Globalization.CultureInfo.InvariantCulture),
            "-t",  duration.ToString("F1", System.Globalization.CultureInfo.InvariantCulture),
            "-i", path,
            "-ac", "1", "-ar", "22050",
            "-f", "wav", "-y", tmp
        }, 45);
        if (!result.Success || !File.Exists(tmp)) return null;
        return tmp;
    }

    public static string? FindBinary(string name)
    {
        var bundleDir = AppContext.BaseDirectory;
        var bundled   = Path.Combine(bundleDir, $"{name}.exe");
        if (File.Exists(bundled)) return bundled;

        var plain = Path.Combine(bundleDir, name);
        if (File.Exists(plain)) return plain;

        // Search PATH
        var pathEnv = Environment.GetEnvironmentVariable("PATH") ?? "";
        foreach (var dir in pathEnv.Split(Path.PathSeparator))
        {
            var full = Path.Combine(dir, $"{name}.exe");
            if (File.Exists(full)) return full;
            var noExt = Path.Combine(dir, name);
            if (File.Exists(noExt)) return noExt;
        }

        // Common install locations
        var commonPaths = new[]
        {
            @"C:\ffmpeg\bin",
            @"C:\Program Files\ffmpeg\bin",
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "ffmpeg", "bin"),
        };
        foreach (var dir in commonPaths)
        {
            var full = Path.Combine(dir, $"{name}.exe");
            if (File.Exists(full)) return full;
        }

        return null;
    }

    public static ProcessResult RunProcess(string binary, string[] args, int timeoutSec)
    {
        var bin = FindBinary(binary);
        if (bin == null)
        {
            Log.Warn($"{binary} not found");
            return new ProcessResult(false, "", $"{binary} not found", Array.Empty<byte>());
        }

        var psi = new System.Diagnostics.ProcessStartInfo(bin)
        {
            UseShellExecute        = false,
            RedirectStandardOutput = true,
            RedirectStandardError  = true,
            CreateNoWindow         = true,
        };
        foreach (var arg in args) psi.ArgumentList.Add(arg);

        using var proc = new System.Diagnostics.Process { StartInfo = psi };
        var stdoutData = new System.Collections.Concurrent.ConcurrentBag<byte[]>();
        var stderrData = new System.Collections.Concurrent.ConcurrentBag<byte[]>();

        proc.OutputDataReceived += (_, e) => { if (e.Data != null) stdoutData.Add(System.Text.Encoding.UTF8.GetBytes(e.Data + "\n")); };
        proc.ErrorDataReceived  += (_, e) => { if (e.Data != null) stderrData.Add(System.Text.Encoding.UTF8.GetBytes(e.Data + "\n")); };

        try { proc.Start(); } catch (Exception ex) { return new ProcessResult(false, "", ex.Message, Array.Empty<byte>()); }

        // For binary stdout (waveform), don't use async reading
        byte[] rawOut;
        if (args.Length > 0 && args[^1] == "-")
        {
            // Raw binary output to stdout
            using var ms = new MemoryStream();
            proc.StandardOutput.BaseStream.CopyTo(ms);
            rawOut = ms.ToArray();
        }
        else
        {
            proc.BeginOutputReadLine();
            proc.BeginErrorReadLine();
            rawOut = Array.Empty<byte>();
        }

        var finished = proc.WaitForExit(timeoutSec * 1000);
        if (!finished) { proc.Kill(true); return new ProcessResult(false, "", $"{binary} timed out", Array.Empty<byte>()); }

        var stdout = rawOut.Length > 0
            ? System.Text.Encoding.UTF8.GetString(rawOut)
            : string.Concat(stdoutData.Select(b => System.Text.Encoding.UTF8.GetString(b)));
        var stderr = string.Concat(stderrData.Select(b => System.Text.Encoding.UTF8.GetString(b)));

        return new ProcessResult(proc.ExitCode == 0, stdout, stderr, rawOut);
    }
}

public record ProcessResult(bool Success, string Stdout, string Stderr, byte[] RawOutput);
